refactor(compiler): extra diagnostics for `@defer` in local compilation mode

AndrewKushnir opened the pull request:
This commit adds extra logic to produce a diagnostic in case `@Component.deferredImports` contain types from imports that also bring eager symbols. This would result in retaining a regular import and generating a dynamic import, which would not allow to defer-load dependencies.

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
refactor(compiler): extra diagnostics for `@defer` in local compilation mode

This commit adds extra logic to produce a diagnostic in case `@Component.deferredImports` contain types from imports that also bring eager symbols. This would result in retaining a regular import and generating a dynamic import, which would not allow to defer-load dependencies.

pmvald commented on packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:
```diff
@@ -656,8 +666,41 @@ export class ComponentDecoratorHandler implements
+    // Check if there are some import declarations that contain symbols used within
+    // the `@Component.deferredImports` field, but those imports contain other symbols
+    // and thus the declaration can not be removed.
+    const nonRemovableImports = this.deferredSymbolTracker.getNonRemovableDeferredImports(context);
```
This diagnostic is file-level, while we are doing it per trait here. Does this mean we may get multiple copies of this diagnostic ??. For example if a file has one bad import and four components in it then we will report the bad import four times (per component) ?? So if this is the case then we can either:

1) modify `getNonRemovableDeferredImports` to narrow down the imports to the ones containing defer imports from the arg `node`. This has the the extra advantage that we can include the `node.name` in the error message and saying something like `This import contains symbols used in the \`@Component.deferredImports\` array of the component ${node.name.getText()}` which further clarifies the issue

2) modify `getNonRemovableDeferredImports` to return a set of bad imports globally for all fikes (and thus takes no arg), and we move this check to `TraitCompiler.resolve` phase after the class/trait loop is done.
3) any other trick to dedup the diagnostic

AndrewKushnir replied:
Great point, I've updated the logic to only retrieve non-deferrable imports based on the currently processed class declaration, which avoids duplication.

pmvald commented on packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:
```diff
@@ -656,8 +666,41 @@ export class ComponentDecoratorHandler implements
+        const diagnostic = makeDiagnostic(
+            ErrorCode.DEFERRED_DEPENDENCY_IMPORTED_EAGERLY, importDecl,
+            `This import contains symbols used in the \`@Component.deferredImports\` array, ` +
+                `but also some other symbols, which prevents Angular compiler from ` +
```
nit; we can also say: ... but also some other symbols that are not in any \`@Component.deferredImports\` array. This renders all these defer imports useless as this import remains and its module is eagerly loaded. To fix this ...

AndrewKushnir replied:
Updated 👍

pmvald commented on packages/compiler-cli/src/ngtsc/annotations/component/src/handler.ts:
```diff
@@ -1193,14 +1233,13 @@ export class ComponentDecoratorHandler implements
   /**
    * Collects a list of deferrable symbols based on the `@Component.deferredImports` field.
```
I think here is a good time to mention this method also registers the explicitly deferred imports (into the deferredSymbolTracker). So that we know it comes with a side effect as well. Optionally this method can be called collectAndRegisterExplicitlyDeferredSymbols. But I imagine it's a way too long name!

AndrewKushnir replied:
Thanks for the comment! I've refactored the code to only extract symbols (and related info) and registration now happens at a different place (at call site).

pmvald commented on packages/compiler-cli/test/ngtsc/local_compilation_spec.ts:
```diff
@@ -1358,6 +1367,164 @@ runInEachFileSystem(() => {
+      it('should support importing multiple deferrable deps from a single file ' +
+             'and use them within `@Component.deferrableImports` field',
+         () => {
+           env.write('deferred-deps.ts', `
```
Do we need to bother with this file? In local compilation we can just import components from any external file as well.

AndrewKushnir replied:
I wanted to test the case when 2 components use symbols from a single import (to make sure we correctly process this case and remove an import afterwards).

AndrewKushnir committed:
fixup! refactor(compiler): extra diagnostics for `@defer` in local compilation mode

AndrewKushnir commented:
@pmvald thanks for the review 👍 I've addressed comments in a fixup commit (https://github.com/angular/angular/pull/53899/commits/bed154aafd8914a06bf9c186d9bcc4c09d64b060).

AndrewKushnir commented:
[Presubmit](http://test/OCL:599015000:BASE:599015021:1705457553559:1a4933c7).

dylhunn commented:
This PR was merged into the repository by commit b0e0f0014ec0e41bd86f8dfadf9a96e67a1c40c8.
